File: chess_backend/test/repertoire_gate.test.js
const test = require('node:test');
const assert = require('node:assert/strict');
const { Chess } = require('chess.js');

const { frontier, gateMoves } = require('../services/repertoireFrontier');
const { tree, drillBranches } = require('../services/repertoireLine');
const {
  createRepertoire, setGate, fenKey,
} = require('../services/repertoireService');

/// The Italian after 1.e4 e5 2.Nf3 Nc6 3.Bc4 Bc5 — White to move, and the
/// position this whole feature came from. Two repertoires start here: one plays
/// 4.b4 (the Evans), the other 4.0-0, and until the gate existed the second
/// one's tree showed the first one's opening.
const ITALIAN = ['e2e4', 'e7e5', 'g1f3', 'b8c6', 'f1c4', 'f8c5'];

function after(...ucis) {
  const board = new Chess();
  for (const uci of ucis) {
    board.move({ from: uci.slice(0, 2), to: uci.slice(2, 4) });
  }
  return board.fen();
}

function keyAfter(...ucis) {
  return fenKey(after(...ucis));
}

const ROOT = after(...ITALIAN);
const ROOT_KEY = fenKey(ROOT);

/// Both openings, kept in one graph — which is where they belong, and the whole
/// reason the view needs a gate.
const BOTH = {
  moves: [
    { fen_key: ROOT_KEY, uci: 'b2b4', san: 'b4', role: 'primary', source: 'chosen' },
    { fen_key: ROOT_KEY, uci: 'e1g1', san: 'O-O', san_alt: '0-0', role: 'alternate', source: 'chosen' },
    // One decision inside each opening, so each has something below it.
    {
      fen_key: keyAfter(...ITALIAN, 'b2b4', 'c5b4'),
      uci: 'c2c3', san: 'c3', role: 'primary', source: 'chosen',
    },
    {
      fen_key: keyAfter(...ITALIAN, 'e1g1', 'g8f6'),
      uci: 'd2d3', san: 'd3', role: 'primary', source: 'chosen',
    },
  ],
  replies: [
    {
      fen_key: keyAfter(...ITALIAN, 'b2b4'),
      uci: 'c5b4', san: 'Bxb4', games: 800, share: '0.80000',
    },
    {
      fen_key: keyAfter(...ITALIAN, 'e1g1'),
      uci: 'g8f6', san: 'Nf6', games: 700, share: '0.70000',
    },
  ],
};

function stubPool({ moves = [], replies = [], skips = [], reviews = [] } = {}) {
  return {
    query: async (text, params) => {
      const flat = text.replace(/\s+/g, ' ').trim();
      const rows = (() => {
        if (flat.includes('SELECT fen_key, uci, san, role')) return moves;
        if (flat.includes('FROM repertoire_skips')) {
          return skips.map((fen_key) => ({ fen_key }));
        }
        if (flat.includes('FROM opening_replies')) {
          const [band, keys] = params;
          return replies.filter(
            (r) => Number(r.min_rating ?? 0) === band && keys.includes(r.fen_key));
        }
        if (flat.includes('SELECT fen_key, due_at, repetitions')) {
          const within = params[2];
          return reviews.filter((row) => within.includes(row.fen_key));
        }
        throw new Error(`Neočekivan upit: ${flat}`);
      })();
      return { rows, rowCount: rows.length };
    },
  };
}

test('the gate narrows the root and touches nothing else', () => {
  // One function, applied to the map the whole walk reads. That is why one
  // filter reaches the queue, the picture, the map and the drill at once.
  const kept = new Map([
    [ROOT_KEY, [{ uci: 'b2b4' }, { uci: 'e1g1' }]],
    ['other', [{ uci: 'd2d4' }]],
  ]);
  gateMoves(kept, ROOT_KEY, 'e1g1');

  assert.deepEqual(kept.get(ROOT_KEY).map((m) => m.uci), ['e1g1']);
  assert.deepEqual(kept.get('other').map((m) => m.uci), ['d2d4']);
});

test('a gate whose move is not kept leaves the position undecided', () => {
  // Rather than leaving it untouched, which would silently show the other
  // opening. "You have not decided this position yet" is exactly true of a
  // repertoire whose first move has not been kept.
  const kept = new Map([[ROOT_KEY, [{ uci: 'b2b4' }]]]);
  gateMoves(kept, ROOT_KEY, 'e1g1');

  assert.deepEqual(kept.get(ROOT_KEY), []);
});

test('the tree draws one opening, not both', async () => {
  // The owner's own words: while working on the 0-0 repertoire, 4.b4 and
  // everything under it must not be on the picture.
  const pool = stubPool(BOTH);
  const drawn = await tree(pool, 7, {
    color: 'w', rootFen: ROOT, gateUci: 'e1g1',
  });

  assert.deepEqual(drawn.children.map((c) => c.uci), ['e1g1']);
  // And the reply under it is still there — the gate narrows the root, not the
  // depth.
  assert.deepEqual(drawn.children[0].children.map((c) => c.san), ['Nf6']);
});

test('without a gate the tree is what it always was', async () => {
  // Every repertoire made before this column has no gate, and must not change.
  const pool = stubPool(BOTH);
  const drawn = await tree(pool, 7, { color: 'w', rootFen: ROOT });

  assert.deepEqual(drawn.children.map((c) => c.uci).sort(),
    ['b2b4', 'e1g1'].sort());
});

test('the queue asks only about the gated opening', async () => {
  // The frontier is what the build screen walks. With the gate on 0-0, no
  // position under 4.b4 may be in it — that is the "clean view" the whole
  // feature is for.
  const pool = stubPool(BOTH);
  const walk = await frontier(pool, 7, {
    color: 'w', rootFen: ROOT, gateUci: 'e1g1',
  });

  const reached = [...walk.open, ...walk.pruned].map((node) => node.path.join(' '));
  assert.equal(reached.some((line) => line.startsWith('b4')), false,
    'red sadrži poziciju iz druge grane');
  assert.ok(walk.branches.every((branch) => branch.path[0] === 'O-O'),
    'tabla grana sadrži granu koja ne ide kroz kapiju');
});

test('the branch list is one opening too', async () => {
  const pool = stubPool(BOTH);
  const listed = await drillBranches(pool, 7, {
    color: 'w', rootFen: ROOT, gateUci: 'e1g1',
  });

  assert.deepEqual(listed.branches.map((b) => b.san), ['O-O Nf6']);
});

/// A pool for the two service functions that write the gate.
function writePool({ rows = [], onWrite = () => {} } = {}) {
  return {
    written: [],
    query: async function query(text, params) {
      const flat = text.replace(/\s+/g, ' ').trim();
      if (flat.startsWith('SELECT root_fen FROM repertoires')) {
        return { rows, rowCount: rows.length };
      }
      if (flat.startsWith('UPDATE repertoires')) {
        onWrite(params);
        return { rows: [{ id: params[0], via_uci: params[2] }], rowCount: 1 };
      }
      if (flat.startsWith('INSERT INTO repertoires')) {
        onWrite(params);
        return {
          rows: [{
            id: 1,
            name: params[1],
            color: params[2],
            root_fen: params[3],
            root_path: params[4],
            via_uci: params[5],
          }],
          rowCount: 1,
        };
      }
      throw new Error(`Neočekivan upit: ${flat}`);
    },
  };
}

test('a gate that cannot be played in that position is refused', async () => {
  // Loud, not stored. A gate that matches nothing is a repertoire that opens on
  // an empty tree with no sentence anywhere saying why — the silent failure
  // this codebase keeps meeting.
  const pool = writePool();
  await assert.rejects(
    () => createRepertoire(pool, 7, {
      name: 'Italijanka', color: 'w', rootFen: ROOT, viaUci: 'a1a8',
    }),
    RangeError,
  );
});

test('a legal gate is stored, and comes back as a move you can read',
  async () => {
    const written = [];
    const pool = writePool({ onWrite: (params) => written.push(params) });
    const made = await createRepertoire(pool, 7, {
      name: 'Italijanka — rokada', color: 'w', rootFen: ROOT, viaUci: 'e1g1',
    });

    assert.equal(made.viaUci, 'e1g1');
    // "O-O", not "e1g1": the server has the board, the client would need one.
    assert.equal(made.viaSan, 'O-O');
    assert.equal(written[0][5], 'e1g1');
  });

test('the gate can be set on a repertoire that already exists, and cleared',
  async () => {
    // The repertoires that most need a gate are the ones built before the
    // column was.
    const pool = writePool({ rows: [{ root_fen: ROOT }] });
    const set = await setGate(pool, 7, { id: 4, viaUci: 'b2b4' });
    assert.equal(set.viaUci, 'b2b4');
    assert.equal(set.viaSan, 'b4');

    const cleared = await setGate(pool, 7, { id: 4, viaUci: null });
    assert.equal(cleared.viaUci, null);
    assert.equal(cleared.viaSan, null);
  });

test('a repertoire that is not yours cannot be gated', async () => {
  const pool = writePool({ rows: [] });
  await assert.rejects(() => setGate(pool, 7, { id: 4, viaUci: 'b2b4' }),
    RangeError);
});

File: chess_app/test/repertoire_gate_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:shared_preferences/shared_preferences.dart';

import 'package:chess_app/features/analysis_studio/services/opening_judge_service.dart';
import 'package:chess_app/features/repertoire/screens/repertoire_build_screen.dart';
import 'package:chess_app/features/repertoire/screens/repertoire_list_screen.dart';
import 'package:chess_app/features/repertoire/services/repertoire_api_service.dart';
import 'package:chess_app/features/repertoire/widgets/repertoire_gate_picker.dart';
import 'package:chess_app/models/analysis_models.dart';

/// The Italian after 1.e4 e5 2.Nf3 Nc6 3.Bc4 Bc5 — White to move, and the
/// position this feature came from: one repertoire plays 4.b4 (the Evans), the
/// other 4.0-0, and until the gate existed each showed the other's opening.
const italian =
    'r1bqk1nr/pppp1ppp/2n5/2b1p3/2B1P3/5N2/PPPP1PPP/RNBQK2R w KQkq - 4 4';

/// After 4.0-0, Black to move.
const afterCastles =
    'r1bqk1nr/pppp1ppp/2n5/2b1p3/2B1P3/5N2/PPPP1PPP/RNBQ1RK1 b kq - 5 4';

/// After 4.0-0 Nf6 — White to move again, a position the screen can ask about.
const afterNf6 =
    'r1bqk2r/pppp1ppp/2n2n2/2b1p3/2B1P3/5N2/PPPP1PPP/RNBQ1RK1 w kq - 6 5';

class _FakeApi extends RepertoireApiService {
  _FakeApi() : super(client: MockClient((_) async => http.Response('{}', 500)));

  /// What each read was asked to narrow itself to.
  final List<String?> treeGates = [];
  final List<String?> frontierGates = [];

  /// What `setGate` was told, and what the root already holds.
  final List<({int id, String? viaUci})> gated = [];
  List<RepertoireMove> keptAtRoot = const [
    RepertoireMove(uci: 'b2b4', san: 'b4', role: 'primary'),
  ];

  @override
  Future<List<RepertoireMove>> movesAt({
    required String color,
    required String fen,
  }) async =>
      fen == italian ? keptAtRoot : const [];

  @override
  Future<RepertoireFrontier?> frontier({
    required String color,
    required String rootFen,
    List<String> rootPath = const [],
    int? minRating,
    String? gateUci,
    String? breadth,
  }) async {
    frontierGates.add(gateUci);
    return const RepertoireFrontier(
      open: [
        FrontierNode(
            fen: afterNf6, path: ['O-O', 'Nf6'], reach: 1, kind: 'undecided')
      ],
      decided: 1,
    );
  }

  @override
  Future<RepertoireTree?> repertoireTree({
    required String color,
    required String rootFen,
    List<String> rootPath = const [],
    int? minRating,
    int maxPly = 16,
    String? gateUci,
    String? breadth,
    List<String> alongPath = const [],
  }) async {
    treeGates.add(gateUci);
    // The server does the narrowing; the fake answers with what a gated read
    // would come back with, so the screen is tested on what it draws.
    return const RepertoireTree(
      rootFen: italian,
      rootPath: ['e4', 'e5', 'Nf3', 'Nc6', 'Bc4', 'Bc5'],
      children: [
        RepertoireTreeMove(
          uci: 'e1g1',
          san: 'O-O',
          fen: afterCastles,
          mine: true,
          role: 'primary',
          children: [
            RepertoireTreeMove(
              uci: 'g8f6',
              san: 'Nf6',
              fen: afterNf6,
              mine: false,
              share: 0.7,
              state: 'open',
            ),
          ],
        ),
      ],
    );
  }

  @override
  Future<Map<String, RepertoireNote>> notes({required String color}) async =>
      const {};

  @override
  Future<Map<String, RepertoireComment>> comments(
          {required String color}) async =>
      const {};

  @override
  Future<List<RepertoireSummary>> list() async => const [
        RepertoireSummary(
          id: 3,
          name: 'Italijanka — rokada',
          color: 'w',
          rootFen: italian,
          moves: 40,
          viaUci: 'e1g1',
          viaSan: 'O-O',
        ),
      ];

  @override
  Future<({bool saved, String? viaUci, String? viaSan})> setGate(
    int id, {
    String? viaUci,
  }) async {
    gated.add((id: id, viaUci: viaUci));
    return (saved: true, viaUci: viaUci, viaSan: viaUci == null ? null : 'O-O');
  }
}

class _SilentJudge implements OpeningJudgeService {
  @override
  bool get hasPersonalToken => false;

  @override
  Future<OpeningJudgeLookup> judge(String fen, String move,
          {int? minRating}) async =>
      const OpeningJudgeLookup.unavailable('no-token');

  @override
  Future<OpponentRepliesLookup> replies(String fen, {int? minRating}) async =>
      const OpponentRepliesLookup.unavailable('no-token');

  @override
  void clearCache() {}
}

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();
  setUp(() => SharedPreferences.setMockInitialValues({}));

  group('the moves offered as a gate', () {
    test('every legal move is there, the kept ones first and marked', () {
      // A repertoire being created may go through a move nobody has kept yet —
      // which is exactly the case that made the gate necessary: 4.0-0 does not
      // exist in the graph until the second repertoire keeps it.
      final options = gateOptionsFor(italian, kept: const ['b2b4']);

      expect(options.first.uci, 'b2b4');
      expect(options.first.kept, isTrue);
      expect(options.any((o) => o.uci == 'e1g1'), isTrue,
          reason: 'rokada nije ponuđena kao kapija');
      // Read from the library, not assembled: castling has three spellings in
      // this codebase and a wrong one would pick a different move.
      expect(options.firstWhere((o) => o.uci == 'e1g1').san, 'O-O');
      expect(options.where((o) => o.kept).length, 1);
    });

    test('a position that cannot be read offers nothing rather than throwing',
        () {
      expect(gateOptionsFor('not a fen'), isEmpty);
    });
  });

  group('the build screen', () {
    testWidgets('carries its gate into every read that walks the graph',
        (tester) async {
      // The tree, the queue and the counts are all one walk. If the gate does
      // not reach them, the "clean view" is a sentence on screen and nothing
      // more.
      final api = _FakeApi();
      await tester.pumpWidget(MaterialApp(
        home: RepertoireBuildScreen(
          name: 'Italijanka — rokada',
          color: 'w',
          rootFen: italian,
          rootPath: const ['e4', 'e5', 'Nf3', 'Nc6', 'Bc4', 'Bc5'],
          gateUci: 'e1g1',
          api: api,
          judge: _SilentJudge(),
          analyse: (fen, depth, multiPV) async => const <AnalysisLine>[],
        ),
      ));
      await tester.pumpAndSettle();

      expect(api.frontierGates, isNotEmpty);
      expect(api.frontierGates.every((gate) => gate == 'e1g1'), isTrue);
      expect(api.treeGates, isNotEmpty);
      expect(api.treeGates.every((gate) => gate == 'e1g1'), isTrue);
    });

    testWidgets('says which opening it is showing', (tester) async {
      // A filtered view that does not say it is filtered is how somebody
      // concludes their work has been deleted.
      final api = _FakeApi();
      await tester.pumpWidget(MaterialApp(
        home: RepertoireBuildScreen(
          name: 'Italijanka — rokada',
          color: 'w',
          rootFen: italian,
          gateUci: 'e1g1',
          api: api,
          judge: _SilentJudge(),
          analyse: (fen, depth, multiPV) async => const <AnalysisLine>[],
        ),
      ));
      await tester.pumpAndSettle();

      expect(find.textContaining('ide kroz O-O'), findsOneWidget);
    });

    testWidgets('without a gate it says nothing and asks for everything',
        (tester) async {
      final api = _FakeApi();
      await tester.pumpWidget(MaterialApp(
        home: RepertoireBuildScreen(
          name: 'Italijanka',
          color: 'w',
          rootFen: italian,
          api: api,
          judge: _SilentJudge(),
          analyse: (fen, depth, multiPV) async => const <AnalysisLine>[],
        ),
      ));
      await tester.pumpAndSettle();

      expect(find.textContaining('ide kroz'), findsNothing);
      expect(api.treeGates.every((gate) => gate == null), isTrue);
    });
  });

  group('the list screen', () {
    testWidgets('names the gate on the card', (tester) async {
      // Two repertoires from one position are otherwise two identical rows with
      // different names.
      final api = _FakeApi();
      await tester
          .pumpWidget(MaterialApp(home: RepertoireListScreen(api: api)));
      await tester.pumpAndSettle();

      expect(find.textContaining('kroz O-O'), findsOneWidget);
    });

    testWidgets('can set the gate on a repertoire that already exists',
        (tester) async {
      // The repertoires that most need a gate are the ones built before the
      // column was — the owner has two of them right now.
      final api = _FakeApi();
      await tester
          .pumpWidget(MaterialApp(home: RepertoireListScreen(api: api)));
      await tester.pumpAndSettle();

      await tester.tap(find.byTooltip('Još'));
      await tester.pumpAndSettle();
      await tester.tap(find.text('Kroz koji potez ide'));
      await tester.pumpAndSettle();

      // The move already kept in that position is at the top, marked.
      expect(find.text('b4'), findsOneWidget);
      expect(find.text('Bez ograničenja'), findsOneWidget);
      await tester.tap(find.text('b4'));
      await tester.pumpAndSettle();

      expect(api.gated.single.id, 3);
      expect(api.gated.single.viaUci, 'b2b4');
    });

    testWidgets('clearing the gate is a decision, closing the sheet is not',
        (tester) async {
      final api = _FakeApi();
      await tester
          .pumpWidget(MaterialApp(home: RepertoireListScreen(api: api)));
      await tester.pumpAndSettle();

      await tester.tap(find.byTooltip('Još'));
      await tester.pumpAndSettle();
      await tester.tap(find.text('Kroz koji potez ide'));
      await tester.pumpAndSettle();
      await tester.tap(find.text('Bez ograničenja'));
      await tester.pumpAndSettle();

      expect(api.gated.single.viaUci, isNull);
    });
  });
}
